parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PCMetrics.java

```diff
@@ -0,0 +1,125 @@
+package io.confluent.parallelconsumer;
+
+/*-
+ * Copyright (C) 2020-2022 Confluent, Inc.
+ */
+
+import io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor;
+import io.confluent.parallelconsumer.internal.State;
+import io.confluent.parallelconsumer.offsets.OffsetEncoding;
+import io.confluent.parallelconsumer.state.ShardKey;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.Timer;
+import lombok.Builder;
+import lombok.Value;
+import lombok.experimental.SuperBuilder;
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Metrics model for Parallel Consumer
+ *
+ * @author Antony Stubbs
+ */
+@Value
+@SuperBuilder(toBuilder = true)
+public class PCMetrics {
```

Comment:
Debug-level metrics like we have in Kafka Streams could be a good approach. Also, MicroMeter allows filtering when setting up the registry so developers can selectively drop metrics. 

```
registry.config()
    .meterFilter(MeterFilter.ignoreTags("too.much.information"))
    .meterFilter(MeterFilter.denyNameStartsWith("jvm"));
```

Reply:
> Also, MicroMeter allows filtering when setting up the registry so developers can selectively drop metrics.

MeterFilter are powerful but annoying. IMHO It would be much more pleasant to be able to enable "basics" and "detailed" metrics using specifics binder or by configuring the binder. One can still use MeterFilters in specific circumstance but they shouldn't be required by the happy path. 

I think per shard metrics are not possible with key level parallelism and a non trivial amount of keys.

1.  IIRC the memory footprint of a timer is in the 2..8kB range depending if percentile / SLO / histogram are enabled. It's likely bearable for most use case but not unremarkable.
2. I think that most TSDB won't be able to handle such cardinality, at least in a cost effective way. 

I wouldn't enable them by default but look for a way to be able to spot anomalies using low cardinality meters, logs or any other means.


Per shard metrics become impossible 

 
It would be nice to not have to rely on metrics filtering to opt-out from shard level metrics, I find it annoying and inefficient compared to be just bind desired meters. Splitting PCMetricsTracker or making it configurable may be one way to do it.

Reply:
Thank you so much for input @cykl !

> I think per shard metrics are not possible with key level parallelism and a non trivial amount of keys.

Do you mean all types of shard metrics? or only timers? or ..?

Reply:
Likely both. 

Timers are worst offenders because each timer generates multiple time-series in the TSDB (ex. count, totalTime, maxTime + buckets for histograms etc.) while a gauge or a counter creates only one.

But if you have many shards, even a single gauge with a shard label could be problematic.

It is often recommended to not use user supplied input for tags values and to be mindful of the cardinality.

Micrometer:
> Beware of the potential for tag values coming from user-supplied sources to blow up the cardinality of a metric. You should always carefully normalize and add bounds to user-supplied input. Sometimes, the cause is sneaky. Consider the URI tag for recording HTTP requests on service endpoints. If we do not constrain 404’s to a value like NOT_FOUND, the dimensionality of the metric would grow with each resource that cannot be found.

Prometheus:
> Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

My understanding is that with key parallelism, shardId = key which can be a very large set of values. In this case you will likely disable shard level metrics because they are too expensive. 

vaguely related question: is PC able to deal with a large number of keys (ex. millions) with key level parallelism? My understanding is that you have a per shard memory overhead and offset maps are going to contains many entries possibly exceeding the 4KB limit. Is there any benchmark / feedback available about that?